docs/httpproxy.md

```diff
@@ -750,7 +750,173 @@ spec:
 
 ## HTTPProxy inclusion
 
-TODO(youngnick) document inclusion
+HTTPProxy permits the splitting of a system's configuration into separate HTTPProxy instances using **inclusion**.
+
+Inclusion, as the name implies, allows for one HTTPProxy object to be included in another, optionally with some conditions inherited from the parent.
+Contour reads the inclusion tree and merges the included routes into one big object internally before rendering Envoy config.
+Importantly, the included HTTPProxy objects do not have to be in the same namespace, so this is functionally the same as the delegation feature of the now-deprecated IngressRoute.
+
+Each tree of HTTPProxy starts with a root, the top level object of the configuration for a particular virtual host.
+Each root HTTPProxy defines a `virtualhost` key, which describes properties such as the fully qualified name of the virtual host, TLS configuration, etc.
+
+HTTPProxies included from the root must not contain a virtualhost key.
+Root objects cannot include other roots either transitively or directly.
+This permits the owner of an HTTPProxy root to allow the inclusion of a portion of the route space inside a virtual host, and to allow that route space to be further subdivided with inclusions.
+Because the path is not necessarily used as the only key, the route space can be multi-dimensional.
+
+### Conditions and Inclusion
+
+Like Routes, Inclusion may specify a set of [conditions](#conditions.
+These conditions are added to any conditions on the routes included.
+This process is recursive.
+
+Conditions are sets of individual condition statements, for example `prefix: /blog` is the condition that the matching request's path must start with `/blog`.
+When conditions are combined through inclusion Contour merges the conditions inherited via inclusion with any conditions specified on the route.
+This may result in duplicates, for example two `prefix:` conditions, or two header match conditions with the same name and value.
+To resolve this Contour applies the following logic.
+
+- `prefix:` conditions are concatenated together in the order they were applied from the root object. For example the conditions, `prefix: /api`, `prefix: /v1` becomes a single `prefix: /api/v1` conditions.
+- Repeated identical `header:` conditions (the same `name`, match type, and `invert` values) are deduplicated to a single `header:` condition.
```

Comment:
The `invert` language is abstracted away from users. They only see `exact` or `notexact` to set that flag. 

Headers aren't merged either. We should discuss what should happen in this case. Should there be some sort of merge? Throw an error back to the user? Or just let them all stack ending up in a potential situation where requests won't match since the route is too specific.

Reply:
For now, I think they need to just be there, and we say that tautological conditions are an error.

Identical headers will end up deduplicated though, right?

Reply:
I'll change the invert part, yes.

Reply:
Currently, master the headers are all just smashed together, so if someone duplicates headers, the route must match all the defined headers to route properly.

There's an issue open (https://github.com/projectcontour/contour/issues/1559), which I thought we would define an error if headers were duplicated. Thoughts on that approach? I'm not sure if use-cases exist where a route should match multiple headers of the same key with different values.

Reply:
I meant that they are deduplicated at the end when applied to Envoy - since setting a correctly-formed header is idempotent on a route (I assume). I think that *contradictory* headers should be an error, but exact matches (where the header name, value, and operator are the same)? That doesn't feel like an error to me.

Reply:
In any case, #1559 is probably the place for this discussion. Can we accept that this is the case for now, and get this one in?

Reply:
It's fine to merge, just want to be careful that we don't explain functionality that doesn't exist or could change and we confuse folks right at the beginning.